### ISSUE TITLE
Support for detecting unmaintained dependencies

### ISSUE THREAD

**Issue by tomkerkhove** [renovatebot/renovate#2305]
**What would you like Renovate to be able to do?**
Keep track of all my dependencies and detect dependencies that have not been updated for a long time.

**Describe the solution you'd like**
What "a long time" means would be defined in the Renovate configuration in terms of days, weeks, months.

**Describe alternatives you've considered**
None.

**Additional context**
This would help us determine if our dependencies are still alive and consider alternatives. This can be because a project has been abandonend but also because it's feature complete. In that case the Renovate customers can simply close the issue then.

**Comment by rarkins** [comment]
What would be the format for alerting you - raise one issue per abandoned dependency, like we do already with deprecated npm dependencies?

**Comment by tomkerkhove** [comment]
That would be ideal for me! In the NuGet/.NET space that information isn't there unfortunately so once a dependency is updated something should be scheduled to check if it's still relevant or not.

Of course, this is a big feature so I'm not expecting this soon.

**Comment by tomkerkhove** [comment]
Does NPM deprecate specific versions or packages in general as well?

I'm mainly interested in packages that I depend on that are no longer being updated and thus probably the project/product died.

**Comment by rarkins** [comment]
Yes individual versions get deprecated. Convention is that if the latest package is deprecated then that means the entire package is considered deprecated.

**Comment by rarkins** [comment]
It's actually not that difficult to implement, particularly because the groundwork from npm deprecations is already in place. I think all that's needed is:

1. Datasources (e.g. nuget) need to return a "last modified date" per-package
2. We add a config option called something like "unmaintainedDaysThreshold" which is an integer value in days for when to consider a package "unmaintained"
3. We raise an issue warning about project being unmaintained if config.unmaintainedDaysThreshold is less than the days since the package's last modified date

**Comment by rarkins** [comment]
@hbetts What do you think of this approach as a compliment to existing deprecations logic?

**Comment by tomkerkhove** [comment]
@samneirinck Do we have the information about last modified date for a NuGet package?

**Comment by samneirinck** [comment]
We could use the `commitTimeStamp` (A timestamp of the most recent commit) [source](https://docs.microsoft.com/en-us/nuget/api/catalog-resource#catalog-index).

**Comment by hbetts** [comment]
> What "a long time" means would be defined in the Renovate configuration in terms of days, weeks, months.

> but also because it's feature complete

Thank you @tomkerkhove for bringing that point up. Knowing when something is no longer maintained is _difficult_ for that reason.

> Convention is that if the latest package is deprecated then that means the entire package is considered deprecated.

@tomkerkhove this is a feature I actually requested under an old account (now deleted, which is why it appears as _Ghost_) - https://github.com/renovatebot/renovate/issues/1592

We have several dependencies in our corporate projects that have marked their latest version as deprecated. In those cases we wanted to notify our teams that they should migrate away from the deprecated packages. While we have not _yet_ deployed that feature internally, it meets our needs to inform development teams of the health of their dependencies.

> I'm mainly interested in packages that I depend on that are no longer being updated and thus probably the project/product died.

This is a good point that I hadn't considered before. While some maintainers are really good at deprecating their packages when they are no longer maintained, some maintainers just stopping working on their packages abruptly.

In that case, the deprecation feature in Renovate doesn't help us.

> @hbetts What do you think of this approach as a compliment to existing deprecations logic?

I don't really know. Something is needed to complement the _deprecation_ feature, but relying on last publish or commit seems very arbitrary.

**Comment by tomkerkhove** [comment]
https://devblogs.microsoft.com/nuget/deprecating-packages-on-nuget-org/

Package deprecation is now a first-class feature. Feels like we can now implement this one

**Comment by tomkerkhove** [comment]
Any thoughts on this?

**Comment by tomkerkhove** [comment]
Bump

**Comment by Jolg42** [comment]
@tomkerkhove I think trying https://socket.dev/ might be interesting for you

**Comment by tomkerkhove** [comment]
That would help but given Renovate already manages my dependency updates it would be nice if it also notifies about dead packages.

**Comment by UnleashMayhem** [comment]
Would also be great for my team. 

Currently this feels like a broken link in Renovate, I want to update my packages when they are behind but I am not aware of the deprecated ones, so I am not receiving updates for the replacement.

**Comment by fgreinacher** [comment]
Could we us [`releaseTimestamp`](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/updates/generate.ts#L277) for this?

**Comment by rarkins** [comment]
Yes, I think so

**Comment by Frulfump** [comment]
@RahulGautamSingh should be closed as duplicate and not completed then
